docs/uniquification.md

```diff
@@ -0,0 +1,31 @@
+# Uniquification
+When assembling and compiling large hardware designs, there are often collisions between names of modules. This issue is especially prominent in the case of using generators, where either the user has to ensure that different generator parameters result in circuits with different names, or the system is responsible for generating unique names. Of course, the latter is desirable.
+
+At the magma level, we can rely on the scoping and lifetime inherent to python to resolve ciruit definitions. That is, if we want to reference a circuit `Foo`, python will resolve `Foo` to the most recently defined circuit with the name `Foo` in the current scope. This part of the resolution comes for free. The issue arises when we want to compile to a backend, e.g. verilog or CoreIR, where scopes are effectively flattened. This document describes the concerns associated with performing uniquification for this purpose and our proposed approach.
+
+## Concerns
+
+### Equality
+One of the major components of uniquification is definining the equality operator between two circuits. The desired characteristics of an equality comparator are:
+- Fast: it should be fast to determine whether or not two circuits are the same (relative to the rest of the compilation).
+- Conservative: the comparator (if not "exact") should never return True if the correct result is False - but can always return False if the correct result is True. That is, false negatives are ok, but false positives are not.
+
+The following are potential implementations of equality operators:
+- Node graph isomorphism: This is an undecidable problem is is computationally intractable. However, it is the most accurate representation of the true equality operator.
+- CoreIR/verilog string comparison: This method is computationally tractable but is bottle-necked by the specific backend compilation flows (which we have found to incur a significant cost). They are also fairly conservative, in that many false negatives are possible. For instance, if instances are re-ordered or renamed, then the circuits will *not* be considered to be equal.
```

Comment:
Reordering might be solved by using a consistent sorting mechanism. Renaming seems tough to deal with.